src/components/utils/src/threads/thread_posix.cc

```diff
@@ -261,15 +270,27 @@ void Thread::stop() {
   SDL_DEBUG("Stopped thread #" << handle_ << " \"" << name_ << " \"");
 }
 
-void Thread::join() {
+void Thread::join(const bool force_stop) {
   SDL_AUTO_TRACE();
   DCHECK_OR_RETURN_VOID(!IsCurrentThread());
 
-  stop();
+  if (force_stop) {
+    stop();
+  }
 
   sync_primitives::AutoLock auto_lock(state_lock_);
+
+  if (!force_stop) {  // then wait on going iteration to finish
+    while (do_iteration_ && thread_created_) {
+      state_cond_.Wait(auto_lock);
+    }
+  }
+
+  stopped_ = true;
+  do_iteration_ = true;
   run_cond_.NotifyOne();
-  if (isThreadRunning_) {
+
+  while (do_iteration_ && thread_created_) {
     state_cond_.Wait(auto_lock);
   }
 }
```

Comment:
@anosach-luxoft, fixed in [ce7130e](https://github.com/smartdevicelink/sdl_core/pull/707/commits/ce7130e0fa97a34d5fb143ff0764557cee82eb68)
Regarding NotifyOne() - it is used only with run_cond_. About state_cond_ - it is always used with Broadcast().